src/backend/distributed/planner/function_call_delegation.c

```diff
@@ -449,3 +556,153 @@ ShardPlacementForFunctionColocatedWithReferenceTable(CitusTableCacheEntry *cache
 
 	return (ShardPlacement *) linitial(placementList);
 }
+
+
+/*
+ * Checks to see if the procedure is being executed on a worker after delegated
+ * by the coordinator. If the flag forcePushdown is set, capture the distribution
+ * argument value, to be used by the planner to make sure that function uses only
+ * the colocated shards of the distribution argument.
+ */
+void
+CheckDelegatedFunctionExecution(DistributedPlanningContext *planContext)
+{
+	List *targetList = planContext->query->targetList;
+	FromExpr *joinTree = planContext->query->jointree;
+	FuncExpr *funcExpr = NULL;
+
+	if (list_length(targetList) > 1 || list_length(joinTree->fromlist) > 1)
+	{
+		/* multiple target list or from list items */
+		return;
+	}
+
+	/*
+	 * Look for a function in the SELECT clause.
+	 */
+	if (list_length(targetList) == 1)
+	{
+		TargetEntry *targetEntry = (TargetEntry *) linitial(targetList);
+		if (IsA(targetEntry->expr, FuncExpr))
+		{
+			funcExpr = (FuncExpr *) targetEntry->expr;
+		}
+	}
+
+	/*
+	 * Look for a function in the FROM clause.
+	 */
+	if (list_length(joinTree->fromlist) == 1)
+	{
+		RangeTblRef *reference = linitial(joinTree->fromlist);
+		if (IsA(reference, RangeTblRef))
+		{
+			RangeTblEntry *rtentry = rt_fetch(reference->rtindex,
+											  planContext->query->rtable);
+
+			if (rtentry->rtekind == RTE_FUNCTION && list_length(rtentry->functions) == 1)
+			{
```

Comment:
can we do early returns and add comments here as well?
```c
if (funcExpr == NULL)
{
    if (rtentry->rtekind != RTE_FUNCTION || list_length(rtentry->functions) != 1)
    {
        /* FROM clause does not contain a regular function */
        return NULL;
    }

    RangeTblFunction *rtfunc = (RangeTblFunction *) linitial(rtentry->functions);
    if (IsA(rtfunc->funcexpr, FuncExpr))
	{
        /* RTE is not a regular function (unexpected?) */
        return NULL;
    }

    funcExpr = (FuncExpr *) rtfunc->funcexpr;
}

```

Reply:
Refactored the code.